### DISCUSSION TITLE
Click Commands and App Context

### DISCUSSION THREAD

**Discussion by gg2** [ckan/ckan#6527] [General]
For the project I'm working on, it has a small set of CLI commands beneath our main project plugin. I'm converting the commands from Pylons to Click. One command makes some changes that trigger another plugin's methods -- IPackageController.before_search. When calling the command, I'm providing it the same config file used to run the CKAN site process.

While testing, the command failed with a RuntimeError: "working outside of application context" when before_search on the other plugin  was called. The other plugin checks Flask's global g for certain attributes in its before_search implementation.

I resolved the problem as follows:

```
from flask import Flask

app = Flask(__name__)

#...

@cli.command()
def my_command():
    with app.app_context():
        #...
```

Are there any suggestions for possibly more appropriate ways to avoid Flask/CKAN losing its context when other methods or plugins execute for command-line operations?
Is it often better to use a separate config file than the main site's config? I believe that was how we resolved an issue between plugins for this set of commands when using Pylons -- we provided a separate, Paster-specific config with certain plugins removed.

**Comment by smotornyuk** [comment]
[Here's](https://github.com/ckan/ckan/blob/master/ckan/cli/user.py#L100-L112) an example of how it can be done(generally, it's exactly what you have done, but an "official" version)

```
@click.pass_context
def remove_user(ctx, username):
    ...
    with ctx.meta['flask_app'].test_request_context():
        ...
```

CKAN stores ready-to-use application under `flask_app` key of the click's  `Context.meta`. You only have to add `click.pass_context` decorator in order to obtain this context.

As for config files, it's better to use the same config file for all the interactions with the CKAN instance.
